src/analysis/souffle/tag_transforms.dl

```diff
@@ -153,6 +155,84 @@ mustHaveIntegrityTag(result, DEFAULT_SQL_OWNER, iTag) :-
   allTagTransformPreconditionsMet(op),
   sqlPolicyRuleNameHasResult(ruleName, $AddIntegrityTag(iTag)).
 
+// In the SQL verifier, we have some merge operations that we would like to
+// propagate integrity tags in a non-default fashion (ie, rather than dropping
+// the integrity tags, we'd like them to intersect or union across all direct
+// inputs). To do that, we introduce a tag transform with special behavior:
+// rather than linking to a `SqlPolicyRule` fact, they shall instead link to
+// built-in rules here that modify how the integrity tags propagate across the
+// wrapped merge operation.
+
+#define RAKSHA_INTERNAL_INTERSECT_ITAG_RULE_NAME "##raksha_internal#intersect_itag_rule"
+#define RAKSHA_INTERNAL_UNION_ITAG_RULE_NAME "##raksha_internal#union_itag_rule"
+
+.decl isSpecialITagPropagationRule(ruleName: symbol)
+isSpecialITagPropagationRule(RAKSHA_INTERNAL_INTERSECT_ITAG_RULE_NAME).
+isSpecialITagPropagationRule(RAKSHA_INTERNAL_UNION_ITAG_RULE_NAME).
+
+.decl isMergeOp(mergeOp: Operation)
+isMergeOp(mergeOp) :- isOperation(mergeOp), operationHasOperator(mergeOp, MERGE_OPERATOR).
+
+.decl isModifiedPropagationMergeOp(mergeOp: Operation, ruleName: symbol)
+isModifiedPropagationMergeOp(mergeOp, ruleName) :-
+  isTagTransformOperation(tagTransformOp),
+  tagTransformHasRuleName(tagTransformOp, ruleName),
+  isSpecialITagPropagationRule(ruleName),
+  operationHasOperandAtIndex(tagTransformOp, transformedInput, 0),
+  operationHasResult(mergeOp, transformedInput),
+  isMergeOp(mergeOp).
+
+.decl mergeOperationHasControlInputStartIndex(mergeOp: Operation, controlStartIndex: number)
+mergeOperationHasControlInputStartIndex(mergeOp, controlStartIdx) :-
+  isMergeOp(mergeOp),
+  operationHasAttribute(mergeOp,
+    [CONTROL_INPUT_START_INDEX_NAME_ATTR, $NumberAttributePayload(controlStartIdx)]).
+
+.decl isMergeOperationDirectInput(mergeOp: Operation, ap: AccessPath)
+isMergeOperationDirectInput(mergeOp, inputPath) :-
+  isMergeOp(mergeOp), isAccessPath(inputPath), operationHasOperandAtIndex(mergeOp, inputPath, idx),
+  mergeOperationHasControlInputStartIndex(mergeOp, controlStartIdx),
+  idx < controlStartIdx.
+
+// For the union rule, any itag on a direct input should propagate to the
+// output.
+mustHaveIntegrityTag(result, DEFAULT_SQL_OWNER, iTag) :-
```

Comment:
Just an observation and question. Result would have certain integrity either because of the default propagation rule or because of this special rule.

I guess if we want to have a special rule for `mayHaveTag`, we will need to do something such as excluding the default propagation rule from triggering?

Reply:
The current default behavior is that we draw edges from all inputs to all outputs, but also draw an edge from a fake access path with no integrity tags. This will propagate all confidentiality tags, but block the default propagation of all integrity tags. As such, I think we're doing OK here. These rules act as a sort of generic tag propagation rule overriding the strict default propagation behavior.